R/utils.R

```diff
@@ -130,6 +131,103 @@ check_sufficient_f_args <- function(f) {
   }
 }
 
+#' Convert to function
+#'
+#' @description
+#' `as_slide_computation()` transforms a one-sided formula into a function.
+#' This powers the lambda syntax in packages like purrr.
+#'
+#' This is an extension of `rlang::as_function` that can create functions that
+#' take three arguments. The arugments can be accessed via the idiomatic
+#' `.x`, `.y`, etc, positional references (`..1`, `..2`, etc), and also by
+#' `slide`-specific names.
+#'
+#' @param x A function or formula.
+#'
+#'   If a **function**, it is used as is.
+#'
+#'   If a **formula**, e.g. `~ .x + 2`, it is converted to a function with up
+#'   to three arguments: `.x` (single argument), or `.x` and `.y`
+#'   (two arguments), or `.x`, `.y`, and `.z` (three arguments). The `.`
+#'   placeholder can be used instead of `.x`, `.group_key` can be used in
+#'   place of `.y`, and `.ref_time_value` can be used in place of `.z`. This
+#'   allows you to create very compact anonymous functions (lambdas) with up
+#'   to two inputs. Functions created from formulas have a special class. Use
+#'   `rlang::is_lambda()` to test for it.
+#'
+#'   If a **string**, the function is looked up in `env`. Note that
+#'   this interface is strictly for user convenience because of the
+#'   scoping issues involved. Package developers should avoid
+#'   supplying functions by name and instead supply them by value.
+#'
+#' @param env Environment in which to fetch the function in case `x`
+#'   is a string.
+#' @export
+#' @inheritParams rlang::args_dots_empty
+#' @inheritParams rlang::args_error_context
+#' @examples
+#' f <- as_slide_computation(~ .x + 1)
+#' f(10)
+#'
+#' g <- as_slide_computation(~ -1 * .)
+#' g(4)
+#'
+#' h <- as_slide_computation(~ .x - .group_key)
+#' h(6, 3)
+#'
+#' @importFrom rlang check_dots_empty0 is_function is_quosure eval_tidy call2
+#'  quo_get_env new_function pairlist2 f_env is_environment missing_arg f_rhs
+#'  is_string quo_get_expr is_formula caller_arg caller_env
+as_slide_computation <- function(x,
```

Comment:
Note: This function and documentation are very lightly modified copies of `rlang::as_function`. I'm not sure what kind of attribution we need here.

Reply:
I can't find any authoritative guidance here.  A bit of searching brought [this](https://community.rstudio.com/t/licensing-a-new-package-that-includes-code-from-another-package/4736/2) up.  Looks like the asker has this package [up on CRAN](https://cran.r-project.org/web/packages/knitrProgressBar/index.html), and here's the DESCRIPTION file, which incorporates some of this.  This might be a good model.
- On the DESCRIPTION part of this, "ctb" seems a bit awkward/misfit but the annotations make it a bit better; we should also add similar "dtc" entries for JHU-CSSE and others that we based the data objects on.  Or maybe find better things from the MARC Code List for Relators?
- `file: Copyrights` + `inst/COPYRIGHTS` part
- Putting copyright stuff in source code: probably this also needs to be in docs as they are part of the "Software" according to MIT license.  Maybe you could do add a `@source` tag describing the situation and including rlang's rendered (not the [template](https://cran.r-project.org/web/licenses/MIT)) license.

---

(((
Initial investigations:

Following their references to the more authoritative (R-wise) Writing R Extensions guide, I see these snippets:

> The mandatory ‘Author’ field describes who wrote the package. It is a plain text field intended for human readers, but not for automatic processing (such as extracting the email addresses of all listed contributors: for that use ‘Authors@R’). Note that all significant contributors must be included: if you wrote an R wrapper for the work of others included in the src directory, you are not the sole (and maybe not even the main) author.

> Both ‘Author’ and ‘Maintainer’ fields can be omitted if a suitable ‘Authors@R’ field is given. This field can be used to provide a refined and machine-readable description of the package “authors” (in particular specifying their precise roles), via suitable R code. It should create an object of class "person", by either a call to person or a series of calls (one per “author”) concatenated by c(): see the example DESCRIPTION file above. The roles can include ‘"aut"’ (author) for full authors, ‘"cre"’ (creator) for the package maintainer, and ‘"ctb"’ (contributor) for other contributors, ‘"cph"’ (copyright holder, which should be the legal name for an institution or corporate body), among others. See ?person for more information. Note that no role is assumed by default. Auto-generated package citation information takes advantage of this specification. The ‘Author’ and ‘Maintainer’ fields are auto-generated from it if needed when building[5](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#FOOT5) or installing.

From `?person`
```
     The new scheme also adds the possibility of specifying _roles_
     based on a subset of the MARC Code List for Relators (<URL:
     https://www.loc.gov/marc/relators/relaterm.html>).  When giving
     the roles of persons in the context of authoring R packages, the
     following usage is suggested.

     ‘"aut"’ (Author) Use for full authors who have made substantial
          contributions to the package and should show up in the
          package citation.

     ‘"com"’ (Compiler) Use for persons who collected code (potentially
          in other languages) but did not make further substantial
          contributions to the package.

     ‘"cph"’ (Copyright holder) Use for all copyright holders.  This is
          a legal concept so should use the legal name of an
          institution or corporate body.

     ‘"cre"’ (Creator) Use for the package maintainer.

     ‘"ctb"’ (Contributor) Use for authors who have made smaller
          contributions (such as code patches etc.) but should not show
          up in the package citation.

     ‘"ctr"’ (Contractor) Use for authors who have been contracted to
          write (parts of) the package and hence do not own
          intellectual property.

     ‘"dtc"’ (Data contributor) Use for persons who contributed data
          sets for the package.

     ‘"fnd"’ (Funder) Use for persons or organizations that furnished
          financial support for the development of the package.

     ‘"rev"’ (Reviewer) Use for persons or organizations responsible
          for reviewing (parts of) the package.

     ‘"ths"’ (Thesis advisor) If the package is part of a thesis, use
          for the thesis advisor.

     ‘"trl"’ (Translator) If the R code is a translation from another
          language (typically S), use for the translator to R.
```
None of these seem like an exact match.  Closest might be "ctb" for rlang / rlang authors and "dtc" for JHU-CSSE and other.  But then rlang / its authors would be mixed in with our actual push contributors.  And the [link](https://www.loc.gov/marc/relators/relaterm.html)'s descriptions also don't seem to line up:
```
 Contributor [ctb]
    A person, family or organization responsible for making contributions to the resource. This includes those whose work has been contributed to a larger work, such as an anthology, serial publication, or other compilation of individual works. If a more specific role is available, prefer that, e.g. editor, compiler, illustrator
     UF [Used For] Collaborator
```
)))

Reply:
Thanks for looking into this! Since this doesn't seem straightforward and since we also want to add attributions for other work outside the changes here, I'll plan to make a separate PR for this.

Reply:
https://github.com/cmu-delphi/epiprocess/issues/327